src/core/components/pin-set.js

```diff
@@ -0,0 +1,269 @@
+'use strict'
+
+const multihashes = require('multihashes')
+const CID = require('cids')
+const protobuf = require('protons')
+const fnv1a = require('fnv1a')
+const dagPB = require('ipld-dag-pb')
+const DAGNode = dagPB.DAGNode
+const DAGLink = dagPB.DAGLink
+const varint = require('varint')
+const once = require('once')
+
+const pbSchema = require('./pin.proto')
+
+const emptyKeyHash = 'QmdfTbBqBPQ7VNxZEYEj14VmRuZBkqFbiwReogJgS1zR1n'
+const emptyKey = multihashes.fromB58String(emptyKeyHash)
+const defaultFanout = 256
+const maxItems = 8192
+const pb = protobuf(pbSchema)
+
+function toB58String (hash) {
+  return new CID(hash).toBaseEncodedString()
+}
+
+function readHeader (rootNode) {
+  // rootNode.data should be a buffer of the format:
+  // < varint(headerLength) | header | itemData... >
+  const rootData = rootNode.data
+  const hdrLength = varint.decode(rootData)
+  const vBytes = varint.decode.bytes
+  if (vBytes <= 0) {
+    throw new Error('Invalid Set header length')
+  }
+  if (vBytes + hdrLength > rootData.length) {
+    throw new Error('Impossibly large set header length')
+  }
+  const hdrSlice = rootData.slice(vBytes, hdrLength + vBytes)
+  const header = pb.Set.decode(hdrSlice)
+  if (header.version !== 1) {
+    throw new Error(`Unsupported Set version: ${header.version}`)
+  }
+  if (header.fanout > rootNode.links.length) {
+    throw new Error('Impossibly large fanout')
+  }
+  return {
+    header: header,
+    data: rootData.slice(hdrLength + vBytes)
+  }
+}
+
+function hash (seed, key) {
+  const buf = Buffer.alloc(4)
+  buf.writeUInt32LE(seed, 0)
+  const data = Buffer.concat([
+    buf, Buffer.from(toB58String(key))
+  ])
+  return fnv1a(data.toString('binary'))
+}
+
+exports = module.exports = function (dag) {
+  const pinSet = {
+    // should this be part of `object` API?
```

Comment:
What's your suggestion here?

Reply:
This was @AdamStone's suggestion so I can't speak to his thoughts on it. It does feel like this method is not in the domain of pinset but I think changing the impl to use an existing tree traversal like `ipld.treestream` might suffice. If we do find it valuable enough to separate, I'm not sure whether object or dag is the proper place.

I attempted an implementation with `ipld.treeStream` but surprisingly ran into a heap OOM during a light pin-set stress test below. Should we look further into that? I can provide a repro test case.

https://github.com/ipfs/js-ipfs/pull/1045/files#diff-fb2d6500d5a91459f37fa2dbc49d3908R101

Reply:
@vmx it turns out that `ipld.treeStream` on a root node created by `pinset.storeItems` ends up running until OOM, like there is a cyclical link or something else to cause an infinite. Counter to that, though, `dag._getRecursive` does not encounter such a problem https://github.com/ipfs/js-ipfs/pull/1045/files#diff-cb106b8602ebb9750632d0d9ce3c1848R83.

I've spent some time looking for a root cause and had to dig into ipld but haven't quite pinpointed it. Is it a serious concern I should spend more time looking into?